Remove paginated crud view mixin from auto update rule view

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->
For the most part, there are no intentional changes to how this view functions for users, though you can view the Screenshots section below to see the visible differences. The one exception is this PR stops collecting deleted automatic update rules in a separate temporary table like so:
![Screen Shot 2022-09-26 at 1 54 41 PM](https://user-images.githubusercontent.com/15785053/192430360-e67e014e-11dc-4833-bc3e-f29df9457ffa.png)
No action can be taken on the deleted rule in that table, and there is a confirmation dialog when attempting to delete a rule that ensures the user is aware which rule is being deleted. Based on that it seems unnecessary to collect this information on the frontend.
## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
Tangentially related to https://dimagi-dev.atlassian.net/browse/SAAS-13760

The motivation for this change is the need to add a view that displays historical information for automatic update rule runs. This could be done on a separate page, but that wouldn't make much sense in the context of a "Edit Data" tab to have a page that displays read-only data. It makes more sense to include this historical information in the same space as the existing rules that run daily. With that in mind, I wanted to add the historical table in a separate tab, and I don't believe we can use tabs with HTML that inherits from the base paginated crud view template.

I've moved away from the base paginated crud view in this PR and tried to keep as much code as possible the same. The javascript is virtually all new, modeled closely after the linked domains javascript since a similar approach was taken there.

I thought about moving `DeduplicationRuleListView` off of the base paginated crud view in this PR too since it previously inherited from `AutomaticUpdateRuleListView`, but it felt potentially premature. A base "rule" view class is what initially came to mind, but figured I'd see if others had opinions before making those changes. For now, the `DeduplicationRuleListView` is decoupled from the `AutomaticUpdateRuleListView`, with some duplicate code existing between the two.

I didn't think there was a good way to break up commits into digestible chunks, so the entire refactor exists within the first commit, and a minor file rename in the second.


### Screenshots
Before
![Screen Shot 2022-09-26 at 9 19 36 PM](https://user-images.githubusercontent.com/15785053/192430988-e8c64dbc-a2f6-4aac-8eec-0067b607e6de.png)

After
![Screen Shot 2022-09-26 at 8 55 09 PM](https://user-images.githubusercontent.com/15785053/192430476-421fea79-73d3-4b10-b7c4-0021db6c9f57.png)
NOTE: I'm planning to add an additional tab shortly, but if the single tab view bothers anyone, I can remove that for now.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Locally tested, and plan to QA this. The goal is not to introduce any changes.
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
https://dimagi-dev.atlassian.net/browse/QA-4570

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x]  The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
remove CRUDPaginatedViewMixin from AutomaticUpdateRuleListView

- add new JS and HTML files that handle pagination and crud operations
- use tabs to prepare for adding a new history tab
- decouple DeduplicationRuleListView from AutomaticUpdateRuleListView for now

orangejenny commented on corehq/apps/data_interfaces/views.py:
```diff
@@ -909,13 +863,50 @@ def initial_rule(self):
+            self.domain,
+            self.rule_workflow,
+            active_only=False,
+        ).order_by('name', 'id')
```
Nitpick, but you handled it so nicely on the js side: https://code.djangoproject.com/ticket/6498

gherceg replied:
Offlined with Jenny about this, but after a bit of a dive, I learned that the django ORM does not override however the backing database handles `ORDER BY` in SQL. In our case, it appears that both `LC_COLLATE` or `LC_CTYPE` is `en_US.utf8` which apparently treats upper case equivalent to lowercase. I compared this to a different type in a raw sql query and saw results sorted in case sensitive order. Based on [this code](https://github.com/dimagi/commcare-cloud/blob/ea3086963827e7beae59252667ca7223732b0f3c/src/commcare_cloud/ansible/roles/postgresql_base/tasks/set_up_dbs.yml#L96-L97) in commcare-cloud, all of our db instances should respect this.

orangejenny replied:
thank you for tracking that down, it's good to know

orangejenny reviewed (approved):
Looks good, my comments are all fussy.

orangejenny commented:
> I thought about moving...

Your approach sounds good to me, given that it sounds like you'll be working in this area for a bit and you're diligent about following up on refactoring.

Another idle thought on code reuse: Biyeun and I have chatted about having a reusable table widget in javascript, possibly a knockout component, possibly replacing our use of datatables(!). So keep front end refactoring in mind, too, please.

I think the single tab is fine but the header should move above it, and probably the help text right under the header, too.

gherceg committed:
fix stickler JS warnings

gherceg commented:
Mentioned to Jenny that I'd have to force push to resolve a strange issue with the `list_automatic_update_rules.html` where it was deleted and then renamed in the following commit. That is now resolved and I don't plan to force push anymore.

gherceg commented on corehq/apps/data_interfaces/static/data_interfaces/js/auto_update_rules.js:
```diff
@@ -0,0 +1,133 @@
+                        var updatedRule = RuleViewModel(data.itemData);
+                        self.rules.push(updatedRule);
+                        self.rules().sort(function (first, second) {
+                            var firstName = first.name.toUpperCase();
```
The rules within the observable array weren't observable objects, so didn't need to unwrap the properties before using. With the ko.mapping.fromJS change I will need to do this though.

Sidenote: I'm not really taking advantage of observables in this page at the moment, apart from changes to `paginatedRules` updating the view, but I propagate those changes to `paginatedRules` manually. I'll look into this a bit to see if there is a cleaner way.

gherceg replied:
See 5a48dce7ad74ed55ca663589f942fe040bb617c1

gherceg commented on corehq/apps/data_interfaces/static/data_interfaces/js/auto_update_rules.js:
```diff
@@ -0,0 +1,133 @@
+                        self.rules.remove(rule);
+                        var updatedRule = RuleViewModel(data.itemData);
+                        self.rules.push(updatedRule);
+                        self.rules().sort(function (first, second) {
```
Agreed 👍🏻

gherceg replied:
See 5a48dce7ad74ed55ca663589f942fe040bb617c1

gherceg commented on corehq/apps/data_interfaces/static/data_interfaces/js/auto_update_rules.js:
```diff
@@ -0,0 +1,133 @@
+
+    var RuleViewModel = function (data) {
+        var self = {};
+        self.action_error = null;
```
Definitely not out of scope. I introduced that mix of casing styles, and agree the boilerplate is a bit annoying. The knockout mapping is cleaner so I'll switch to that. If I really wanted to use camelCase exclusively in the frontend, I should update the json format when preparing the list of rules to send to the frontend.

gherceg replied:
See 5a48dce7ad74ed55ca663589f942fe040bb617c1

gherceg commented on corehq/apps/data_interfaces/templates/data_interfaces/auto_update_rules.html:
```diff
@@ -0,0 +1,23 @@
+{% requirejs_main 'data_interfaces/js/auto_update_rules' %}
+
+{% block page_content %}
+  {% initial_page_data 'help_site_url' help_site_url %}
```
That's a good question. The DeduplicationRuleListView used a different template previously, but this variable enabled typing the url once to be used in either template. Now that they are separated it isn't necessary anyway so I'll hardcode that url into the html template directly.

gherceg replied:
See 5a48dce7ad74ed55ca663589f942fe040bb617c1

gherceg committed:
first batch of addressing feedback

gherceg commented on corehq/apps/data_interfaces/templates/data_interfaces/auto_update_rules.html:
```diff
@@ -0,0 +1,23 @@
+
+  <div id="ko-auto-update-rules" class="ko-template">
+    <ul class="nav nav-tabs">
+      <li class="active"><a data-toggle="tab" href="#tabs-update-rules">Automatic Update Rules</a></li>
```
See 5a48dce7ad74ed55ca663589f942fe040bb617c1

gherceg commented on corehq/apps/data_interfaces/templates/data_interfaces/partials/auto_update_rule_list.html:
```diff
@@ -0,0 +1,110 @@
+{% load hq_shared_tags %}
+{% load i18n %}
+
+<div id="ko-tabs-update-rules" data-bind="with: $root">
```
See 5a48dce7ad74ed55ca663589f942fe040bb617c1

gherceg commented on corehq/apps/data_interfaces/static/data_interfaces/js/auto_update_rules.js:
```diff
@@ -0,0 +1,133 @@
+            $.ajax({
+                url: "",
+                type: "POST",
+                dataType: 'json',
```
According to jquery docs, this is the default
> dataType (default: Intelligent Guess (xml, json, script, or html))
So sounds like it would probably work if not specified, but I figure it is easy enough to keep it?

gherceg committed:
second batch of addressing feedback

gherceg commented on corehq/apps/data_interfaces/static/data_interfaces/js/auto_update_rules.js:
```diff
@@ -0,0 +1,133 @@
+
+        self.showActionError = function (rule, error) {
+            var ruleToUpdate = self.rulesById()[rule.id];
+            ruleToUpdate.action_error = error;
```
My understanding is if the user takes a successful action on that rule, it'll clear `action_error` when handling the successful response by basically clobbering the existing rule and replacing it with the rule returned in the response. This seems to be how it behaved previously as well.

orangejenny replied:
cool, sounds reasonable

gherceg commented on corehq/apps/data_interfaces/views.py:
```diff
@@ -642,48 +643,32 @@ class AutomaticUpdateRuleListView(DataInterfaceSection, CRUDPaginatedViewMixin):
+        response = self._update_rule(request.POST['id'], request.POST['action'])
+        return JsonResponse(response)
+
+    def page_url(self):
```
Eh just an artifact of the view I was using as a template. It seems to work just fine. See d9cce87be6202fcddc5dcc8a7887e91669aced1b

gherceg commented on corehq/apps/data_interfaces/views.py:
```diff
@@ -642,48 +643,32 @@ class AutomaticUpdateRuleListView(DataInterfaceSection, CRUDPaginatedViewMixin):
-        context = self.pagination_context
-        domain_obj = Domain.get_by_name(self.domain)
-        hour = domain_obj.auto_case_update_hour
+    def get(self, request, *args, **kwargs):
```
Misunderstanding of how django views worked on my part. See d9cce87be6202fcddc5dcc8a7887e91669aced1b

gherceg commented on corehq/apps/data_interfaces/templates/data_interfaces/partials/auto_update_rule_list.html:
```diff
@@ -0,0 +1,110 @@
+      </table>
+      <pagination data-apply-bindings="false"
+        params="goToPage: goToPage,
+                perPage: itemsPerPage,
```
See d9cce87be6202fcddc5dcc8a7887e91669aced1b

gherceg committed:
ignore stickler syntax error

gherceg committed:
unwrap id in delete rule modal

gherceg commented:
@orangejenny so you are imagining something like this?
![Screen Shot 2022-09-27 at 5 43 34 PM](https://user-images.githubusercontent.com/15785053/192662219-b4b0b91f-d168-4fc1-9844-4e39414445f2.png)

gherceg committed:
move header and help text outside of tabs

orangejenny commented on corehq/apps/data_interfaces/views.py:
```diff
@@ -880,7 +880,7 @@ def page_context(self):
         context.update({
             'help_site_url': 'https://confluence.dimagi.com/display/commcarepublic/Automatically+Close+Cases',
-            'time': f"{hour}:00" if hour else _('midnight'),
+            'time': f"{hour}:00" if hour else _('midnight'),  # noqa: E999
```
I was confused by stickler flagging this earlier - what is it that the linter doesn't like here?

gherceg replied:
I too am confused by it. Daniel mentioned that stickler runs on 3.8, so I tried running this in a 3.8 shell but it works fine there too. I'm assuming it is some combination of the f-string formatting, translation underscore and ternary statement that is throwing stickler off? Would be nice to have a more conclusive answer, but it seems safe to ignore.

orangejenny replied:
huh, mysterious, agree it's fine to ignore

orangejenny commented on corehq/apps/data_interfaces/views.py:
```diff
@@ -651,20 +651,18 @@ class AutomaticUpdateRuleListView(DataInterfaceSection):
-    def get(self, request, *args, **kwargs):
+    @property
+    def page_context(self):
         context = super().main_context
```
Didn't notice this earlier: is it intentional this starts with super's main_context instead of super's page_context? In general you don't need to think about main_context, it just pulls from the `page_title`, etc attributes and ultimately gets [combined with page_context](https://github.com/dimagi/commcare-hq/blob/14793c049a2a2691a02db5660d015166b0b1913e/corehq/apps/hqwebapp/views.py#L902-L906).

I'm being picky because our views are inconsistent about how they handle the page_context/main_context and the navigation attributes, which makes our navigation annoying inconsistent. Reminds me of these improvements Charl just made: https://github.com/dimagi/commcare-hq/pull/32179

gherceg replied:
Again just based on the view I was using as a reference. That makes a lot of sense and thanks for being picky! The inconsistencies only add to the confusion so it is well worth resolving. See 56b14a5513b58cf87db5a21d1ff72e7767f0312a

orangejenny reviewed (commented):
This looks great, thank you for the quick responses. I just have one new question to resolve.

gherceg committed:
use page_context, not main_context

gherceg committed:
rename list_automatic_update_rules.js to deduplication_rules.js

gherceg committed:
duplicate format_rule code into deduplication list view (the irony)

planning to do refactoring the deduplication view in the near
future which will remove duplicate code

gherceg committed:
use get_timezone_for_user directly

gherceg committed:
missing time from page context

gherceg committed:
Merge branch 'master' into gh/replace-base-crud-view

gherceg committed:
add allowed_actions to dedupe view

orangejenny commented on corehq/apps/data_interfaces/views.py:
```diff
@@ -667,19 +667,14 @@ def post(self, request, *args, **kwargs):
         return EditCaseRuleView.urlname
 
-    @property
-    @memoized
```
kind of a shame we don't cache `get_timezone_for_user`, but feel free to call that out of scope

gherceg replied:
Yeah I'm going to pull the out of scope card, though I did create [a ticket](https://dimagi-dev.atlassian.net/browse/SAAS-14137) for it.

orangejenny replied:
that counts as something
